src/main/scala/org/broadinstitute/dsde/workbench/sam/service/PermissionsByUsers.scala

```diff
@@ -0,0 +1,83 @@
+package org.broadinstitute.dsde.workbench.sam.service
+
+import org.broadinstitute.dsde.workbench.model.WorkbenchSubject
+import org.broadinstitute.dsde.workbench.sam.audit.AccessChange
+import org.broadinstitute.dsde.workbench.sam.model.{AccessPolicy, ResourceAction, ResourceRoleName, ResourceTypeName}
+import org.broadinstitute.dsde.workbench.sam.util.groupByFirstInPair
+
+/**
+  * Utility class for comparing collections of policies for use in audit logs
+  * @param policies
+  */
+private[service] class PermissionsByUsers(policies: Iterable[AccessPolicy]) {
+  private val memberRoles: Set[(WorkbenchSubject, ResourceRoleName)] = flattenMemberRoles(policies)
+  private val memberActions: Set[(WorkbenchSubject, ResourceAction)] = flattenMemberActions(policies)
+  private val memberDescendantRoles: Set[(WorkbenchSubject, (ResourceTypeName, ResourceRoleName))] = flattenMemberDependantRoles(policies)
+  private val memberDescendantActions: Set[(WorkbenchSubject, (ResourceTypeName, ResourceAction))] = flattenMemberDependantActions(policies)
+
+  /**
+    * Returns the set of access changes if all the permissions in `other` are removed from `this`.
+    */
+  def removeAll(other: PermissionsByUsers): Set[AccessChange] = {
+    val memberRolesDiff = groupByFirstInPair(this.memberRoles removedAll other.memberRoles)
+    val memberActionsDiff = groupByFirstInPair(this.memberActions removedAll other.memberActions)
+    val memberDescendantRolesDiff = groupByFirstInPair(this.memberDescendantRoles removedAll other.memberDescendantRoles)
+    val memberDescendantActionsDiff = groupByFirstInPair(this.memberDescendantActions removedAll other.memberDescendantActions)
+    compileDiff(memberRolesDiff, memberActionsDiff, memberDescendantRolesDiff, memberDescendantActionsDiff)
+  }
+
+  private def compileDiff(
+      memberRolesDiff: Map[WorkbenchSubject, Iterable[ResourceRoleName]],
+      memberActionsDiff: Map[WorkbenchSubject, Iterable[ResourceAction]],
+      memberDescendantRolesDiff: Map[WorkbenchSubject, Iterable[(ResourceTypeName, ResourceRoleName)]],
+      memberDescendantActionsDiff: Map[WorkbenchSubject, Iterable[(ResourceTypeName, ResourceAction)]]) = {
+    for {
+      member <- memberRolesDiff.keySet ++ memberActionsDiff.keySet ++ memberDescendantRolesDiff.keySet ++ memberDescendantActionsDiff.keySet
+    } yield {
+      AccessChange(
+        member,
+        memberRolesDiff.get(member),
+        memberActionsDiff.get(member),
+        memberDescendantRolesDiff.get(member).map(groupByFirstInPair),
+        memberDescendantActionsDiff.get(member).map(groupByFirstInPair)
+      )
+    }
+  }
+
+  private def flattenMemberRoles(policies: Iterable[AccessPolicy]): Set[(WorkbenchSubject, ResourceRoleName)] =
```

Comment:
These for-comps are nice though I think we could remove two of them by passing a selector in, something like (forgive the psuedo-types)
```scala
def flattenMember[A](f: Policy -> A, policies: Set[Policy]): Set[(WorkbenchSubject, A)]
def flattenMemberDependant[A](f: DependentPermission -> A, policies: Set[Policy]): Set[(WorkbenchSubject, (ResourceTypeName, A))]
```